array.go

```diff
@@ -913,3 +913,40 @@ func countElement(array []interface{}, element interface{}) int {
 	}
 	return count
 }
+
+func (a *Array) Filter(filter func(index int, value *Value) bool) *Array {
+	a.chain.enter("Filter()")
+	defer a.chain.leave()
+
+	if a.chain.failed() {
+		return a
+	}
+
+	if filter == nil {
+		a.chain.fail(AssertionFailure{
+			Type: AssertUsage,
+			Errors: []error{
+				errors.New("unexpected nil function argument"),
+			},
+		})
+	}
+
+	var filteredArray []interface{}
+
+	chainFailure := false
+	for index, element := range a.value {
+		valueChain := a.chain.clone()
+		valueChain.setFailCallback(func() {
+			chainFailure = true
+		})
+		if filter(index, newValue(valueChain, element)) {
+			filteredArray = append(filteredArray, element)
+		}
+	}
+
+	if chainFailure {
+		a.chain.setFailed()
+	}
+	a.value = filteredArray
+	return a
```

Comment:
Instead of modifying the original array, we need to return a new one, which content is filtered.

Reply:
I'm not sure how this should be done, do we have a similar example in the repo? Should I use the newArray method or did you have something else in mind?

Reply:
Yep, you just need to construct a slice with filtered values and pass it to newArray.

Reply:
I'm sorry, but I can't figure it out - every time I use the newArray function it just copies the previous chain with its values unaltered and I have no idea how to make it work without overwriting the old values. Any hints?

Reply:
TBH I don't quite understand the question. Maybe @angshumanHalder could show his code from early versions of #176? That PR had very similar code initially.

@vtech6 or you can show me specific code snippet and show when you have problematic / unexpected behavior.

Reply:
@vtech6 you could do something like this 
```go
	array := make([]interface{}, 0)
	return newArray(a.chain, array)
```

where `a` is your original array

Reply:
@angshumanHalder when I do this, the values of the array stay exactly how they were before the filter, for example when I run the following test: 
```
	        array := NewArray(reporter, []interface{}{1, 2, 3, 4, 5, 6})
		array.Filter(func(index int, value *Value) bool {
			return value.Raw() != 2.0 && value.Raw() != 5.0
		})
		assert.Equal(t, []interface{}{1.0, 3.0, 4.0, 6.0}, array.Raw())
```
I get this error:
```
Not equal: 
        	            	expected: []interface {}{1, 3, 4, 6}
        	            	actual  : []interface {}{1, 2, 3, 4, 5, 6}
```
Here's the implementation of the filter:
```
var filteredArray = []interface{}

for index, element := range a.value {
		valueChain := a.chain.clone()
		valueChain.setFailCallback(func() {
			chainFailure = true
		})

		if filter(index, newValue(valueChain, element)) {
			filteredArray = append(filteredArray, element)
		} else {
			valueChain.replace("Filter[%v]", element)
		}

	}
...

return newArray(a.chain, filteredArray)
```
It works when I directly overwrite the value with `a.value = filteredArray` but it doesn't when I call the function above.

Also maybe you can help me verify if I use the `valueChain.replace` in the correct spot. Thanks for the feedback!

Reply:
I guess that's because in the test above you're checking the original array, and instead you should check the array returned from Filter().

Reply:
Oh, you're absolutely right! Let me take a look.

Reply:
I confirm it works now :) I'll try to commit the issues tomorrow, just want to go through everything again polish it up a little.

Reply:
Great, ping me when it's ready for review.